lib/Core/Memory.cpp

```diff
@@ -630,66 +630,91 @@ void ObjectState::print() const {
 
 /***/
 
-PersistentState::PersistentState(const ObjectState *os, 
-                                 const Array *cacheLines,
-                                 const Array *rootCauses,
-                                 const Array *idxArray)
+PersistentState::PersistentState(const ObjectState *os)
   : ObjectState(*os),
-    cacheLineUpdates(cacheLines, nullptr),
-    pendingCacheLineUpdates(cacheLineUpdates),
-    idxEmptyUpdates(idxArray, nullptr),
-    rootCauseLocations(rootCauses, nullptr) {
-  // -- For cacheline stuff
-  // 0 <= X < 2^32
-  idxUnbounded = ReadExpr::create(idxEmptyUpdates,
+    cacheLineUpdates(nullptr, nullptr),
+    pendingCacheLineUpdates(nullptr, nullptr),
+    rootCauseLocations(nullptr, nullptr),
+    pendingRootCauseLocations(nullptr, nullptr) {
+
+  // Set up all the symbolic Arrays we need.
+  ArrayCache *arrayCache = getArrayCache();
+  const MemoryObject *object = getObject();
+  std::string baseName = object->name;
+  uint64_t size = object->parent->getSizeInCacheLines(object->size);
+
+  // For initializing values
+  std::vector<ref<ConstantExpr> > Init(size);
+
+  // First, the symbolic cache line tracking array (initialize to persisted).
+  Init.assign(size, getPersistedExpr());
+  const Array *cacheLines = arrayCache->CreateArray(baseName + "_cacheLines", size,
+                                                    &Init[0], &Init[0] + size,
+                                                    Expr::Int32 /* domain */,
+                                                    Expr::Int8 /* range */);
+  cacheLineUpdates = UpdateList(cacheLines, nullptr);
+  pendingCacheLineUpdates = UpdateList(cacheLineUpdates);
+
+  // Set up the root causes symbolic array (initialize to nullptr).
+  Init.assign(size, getNullptr());
+  const Array *rootCauses = arrayCache->CreateArray(baseName + "_rootCauses", size,
+                                                    &Init[0], &Init[0] + size,
+                                                    Expr::Int32 /* domain */,
+                                                    Expr::Int64 /* range */);
+  rootCauseLocations = UpdateList(rootCauses, nullptr);
+
+  // Set up a symbolic integer to act as an "arbitrary offset" into this.
+  const Array *idxArray = arrayCache->CreateArray(baseName + "_idx", 1,
+                                                       nullptr, nullptr,
+                                                       Expr::Int32,
+                                                       Expr::Int32);
+  // ReadExpr just deep copies the UpdateList you give it;
```

Comment:
I love fun facts.

Reply:
Oops, it's actually not accurate. It is a copy, but not a deep one. Still makes the same point.

Reply:
Update comment to be accurate